Enhanced <select>: <selectedcontent> implementation

annevk opened the pull request:
#### 6bce16730911644e7455f5d48b3157c95c02f0ab
<pre>
Enhanced &lt;select&gt;: &lt;selectedcontent&gt; implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=301929">https://bugs.webkit.org/show_bug.cgi?id=301929</a>
<a href="https://rdar.apple.com/164005712">rdar://164005712</a>

Reviewed by Ryosuke Niwa.

This follows the latest standardization plan to always update all
&lt;selectedcontent&gt; elements and thereby avoids the issues of an earlier
version that would remove nodes during node removal which is a big
no-no.

For additional hardening it is only enabled when MutationEventsEnabled
is false.

Some tests are failing as specification updates and test changes are
still under review. This will be corrected in subsequent changes.

With all the settings in their current state (i.e., this is not
enabled) this tested as neutral on Speedometer.

Canonical link: <a href="https://commits.webkit.org/307064@main">https://commits.webkit.org/307064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/811407f24599b007e71532438ac9bb40f66849f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143277 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/15752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/6978 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/151953 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/96498 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/6b2ef71c-0d9e-4de2-a5c5-e4c8e1bfc19b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/145144 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/16412 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/15833 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/110192 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/79334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/9d8357d6-6457-46d2-b098-88eb03abfd8b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/146226 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/12633 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/128558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/91102 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/636a45d6-3f27-443e-8bb9-68e9ac859b87) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/12117 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/9835 "Passed tests") | [✅ 🛠 wpe-libwebrtc](https://ews-build.webkit.org/#/builders/172/builds/1951 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/135276 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/121548 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/5021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/154264 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/4089 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/15796 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/169/builds/6014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/118215 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/15833 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/13322 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/118554 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/14487 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/126215 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/71191 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/22085 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/15421 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/4848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/174574 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/15155 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/79141 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/45077 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/15366 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/15217 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1289,6 +1289,16 @@ ExceptionOr<void> ContainerNode::replaceChildren(FixedVector<NodeOrString>&& vec
+    ChildListMutationScope mutation(*this);
+    NodeVector removedChildren;
+    removeAllChildrenWithScriptAssertionMaybeAsync(ChildChange::Source::API, removedChildren, DeferChildrenChanged::No);
+    insertChildrenBeforeWithoutPreInsertionValidityCheck(WTFMove(newChildren));
```
We need to check for exception here.

annevk replied:
There should not be an exception as the new children were the children of another element so that's all taken care of. Added an assert and renamed to `replaceChildrenWithoutValidityCheck` for clarity.

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -122,6 +124,27 @@ void HTMLOptionElement::removedFromAncestor(RemovalType removalType, ContainerNo
+        return;
+
+    if (RefPtr selectedContent = select->enabledSelectedContent()) {
+        ScriptDisallowedScope::EventAllowedScope eventAllowedScope { *selectedContent };
```
Why is this okay? I don't think it is. `cloneIntoSelectedContent` mutates script visible DOM tree which can have mutation events, etc...

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1765,6 +1769,66 @@ ExceptionOr<void> HTMLSelectElement::showPicker()
+        return nullptr;
+    }();
+
+    ScriptDisallowedScope::EventAllowedScope eventAllowedScope { *selectedContent };
```
Ditto. I don't think this is okay.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1765,6 +1769,66 @@ ExceptionOr<void> HTMLSelectElement::showPicker()
+    ScriptDisallowedScope::EventAllowedScope eventAllowedScope { *selectedContent };
+
+    if (!selectedOption)
+        selectedContent->removeChildren();
```
This can fire `unload` event in iframes for example.

rniwa commented on Source/WebCore/html/HTMLSelectedContentElement.cpp:
```diff
@@ -46,4 +50,59 @@ Ref<HTMLSelectedContentElement> HTMLSelectedContentElement::create(const Qualifi
+        return;
+
+    if (RefPtr select = lineageOfType<HTMLSelectElement>(oldParentOfRemovedTree).first()) {
+        select->updateSelectedContent();
```
`removedFromAncestor` isn't supposed to mutate DOM like this.

rniwa reviewed (changes requested):
I don't think none of `ScriptDisallowedScope::EventAllowedScope` introduced in this PR is safe.

rniwa commented on Source/WebCore/html/HTMLOptionElement.cpp:
```diff
@@ -122,6 +124,26 @@ void HTMLOptionElement::removedFromAncestor(RemovalType removalType, ContainerNo
+    if (!select)
+        return;
+
+    select->updateSelectedContent();
```
I'm not sure if it's always safe to mutate DOM in `finishParsingChildren`.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1765,6 +1770,41 @@ ExceptionOr<void> HTMLSelectElement::showPicker()
+    }
+
+    for (Ref selectedContent : selectedContentElements) {
+        ScriptDisallowedScope::EventAllowedScope eventAllowedScope { selectedContent };
```
Why is this safe? I don't think it is.

rniwa replied:
Is this even needed? It doesn't seem like this code is getting triggered inside a `ScriptDisallowedScope`?

annevk replied:
Indeed it's not. I think it was before due to the loop and then I realized the loop was bad but not that moving it out of the loop would no longer require a scope.

nt1m commented on Source/WebCore/html/HTMLSelectElement.h:
```diff
@@ -44,7 +44,6 @@ class ShadowRoot;
 
 #if !PLATFORM(IOS_FAMILY)
 class PopupMenu;
-#endif
```
this removal is causing a build error

rniwa commented on Source/WebCore/html/HTMLSelectedContentElement.cpp:
```diff
@@ -46,4 +50,40 @@ Ref<HTMLSelectedContentElement> HTMLSelectedContentElement::create(const Qualifi
+    RefPtr<HTMLSelectElement> nearestAncestorSelect;
+    m_isDisabled = false;
+    for (Ref ancestor : ancestorsOfType<HTMLElement>(*this)) {
+        if (RefPtr select = dynamicDowncast<HTMLSelectElement>(ancestor)) {
```
I don't think this needs to be a smart pointer. `auto*` should do.

rniwa commented on Source/WebCore/html/HTMLSelectedContentElement.cpp:
```diff
@@ -46,4 +50,40 @@ Ref<HTMLSelectedContentElement> HTMLSelectedContentElement::create(const Qualifi
+void HTMLSelectedContentElement::didFinishInsertingNode()
+{
+    RefPtr<HTMLSelectElement> nearestAncestorSelect;
+    m_isDisabled = false;
```
It seems to me that we want to update `m_isDisabled` synchronously in `HTMLSelectedContentElement::insertedIntoAncestor`.
Otherwise, m_isDisabled can get out of sync with the DOM tree because scripts may run before `didFinishInsertingNode()` is called.
The only thing we want to call in this function is `updateSelectedContent()`.

annevk replied:
So we search for an ancestor `select` element twice? Once in the insertion steps and once in the post-insertion steps? What if we hit any of the invalidating conditions during the post-insertion steps?

annevk replied:
I also feel like I'm missing something about how it can get out of sync as we only manage this state here so it's based on the most recent version of the tree.

rniwa replied:
What I'm saying is that between `insertedIntoAncestor` and `didFinishInsertingNode`, some scripts may run and that script will observe that `m_isDisabled` is `false` even though it should be `true` based on the DOM tree at that moment (since the insertion had already happened).

rniwa replied:
Hm... I guess `isDisabled` isn't exposed to scripts? In that case, maybe the current approach is fine.
One subtle thing is that if there is any script that runs between `insertedIntoAncestor` and `didFinishInsertingNode`, WebKit behaves as if this content element had not been inserted into yet.

annevk replied:
We only really consult the state as part of `HTMLSelectElement::updateSelectedContent()`. I guess if you insert multiple `selectedcontent` elements at once and a later one is technically disabled (by being a child of something), it probably gets updated due to the first post-insertion running `updateSelectedContent()` and the later one not getting filtered out in time.

I'm not sure if that's worse than walking the tree twice in the common case however for which we try to optimize.

rniwa replied:
Hm... so maybe what we want to do is check that disabled-ness as it traverses `descendantsOfType<HTMLSelectedContentElement>`? That way, `m_isDisabled` is updated based on the latest DOM state even if there was nested script executions between `insertedIntoAncestor` and `didFinishInsertingNode`. It's already traversing every `HTMLSelectedContentElement` anyway.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1829,6 +1832,39 @@ ExceptionOr<void> HTMLSelectElement::showPicker()
+    if (m_multiple)
+        return;
+
+    RefPtr selectedOption = [&] -> RefPtr<HTMLOptionElement> {
```
Maybe we can use `std::ranges::find_if`?

annevk replied:
I end up with
```c++
    RefPtr selectedOption = [&] -> RefPtr<HTMLOptionElement> {
        auto it = std::ranges::find_if(listItems(), [](auto& element) {
            if (RefPtr option = dynamicDowncast<HTMLOptionElement>(*element))
                return option->selected();
            return false;
        });
        return it != listItems().end() ? downcast<HTMLOptionElement>(**it) : nullptr;
    }();
```
which seems worse? But I might be holding it wrong.

nt1m replied:
`Vector::findIf` might be cleaner here

annevk replied:
You still end up downcasting twice.

nt1m replied:
Hmm, listItems() should probably return a vector of HTMLOptionElement, then you can write it like this:

```
RefPtr selectedOption = listItems().findIf([]() {
  
});
```

Anyway, this could be a future clean up

rniwa replied:
I think `listItems` does return a `Vector` so maybe this?
```cpp
auto& list = listItems();
auto index = list.findIf([](auto& element) {
    if (RefPtr option = dynamicDowncast<HTMLOptionElement>(*element))
        return option->selected();
    return false;
});
RefPtr selectedOption = index != notFound ? list[index].get() : nullptr;
```

But the current code is fine as is.

annevk replied:
It does return a `Vector`, but it can also contain `HTMLOptGroupElement` and `HTMLHrElement` (which are important for other accessors btw) so you have to `downcast` again if you want to end up with an `HTMLOptionElement`.

rniwa replied:
Oh I see. The current loop is fine then.

rniwa commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1829,6 +1832,39 @@ ExceptionOr<void> HTMLSelectElement::showPicker()
+        if (!selectedOption)
+            selectedContent->removeChildren();
+        else
+            selectedOption->cloneIntoSelectedContent(selectedContent);
```
Don't we want co clear `selectedOption` here? 
Otherwise, wouldn't we end up cloning into multiple selected content elements?

annevk replied:
Possibly cloning into multiple elements is okay. Web developers are only supposed to use one `selectedcontent` element per `select` element, but this is the desired fallback. This also avoids having to update `selectedcontent` elements during their removal for instance.

rniwa replied:
ah, ok so we're changing the spec to do this instead?

annevk replied:
Yeah, there's an issue tracking this and it has agreement, just no PR yet.

rniwa replied:
Hm... so there is a chance between the time we collected `HTMLSelectedContentElement` elements and  `cloneIntoSelectedContent` is called, some scripts will execute as a result of a prior call to `cloneIntoSelectedContent` as that runs post insertion steps synchronously. Is that out-of-state situation fine in terms of the spec?

annevk replied:
I think it is, but I think between this and disabled we should try to think of a cleaner model. I don't think it should hold up this PR however as we have a lot of wiggle room in non-conforming scenarios. For now our main focus is people using a single `selectedcontent` element.

So my plan is to land this PR and open another issue on the HTML standard to discuss when to resolve disabled.

rniwa replied:
Sounds good to me.

annevk committed:
Enhanced <select>: <selectedcontent> implementation
https://bugs.webkit.org/show_bug.cgi?id=301929
rdar://164005712

Reviewed by Ryosuke Niwa.

This follows the latest standardization plan to always update all
<selectedcontent> elements and thereby avoids the issues of an earlier
version that would remove nodes during node removal which is a big
no-no.

For additional hardening it is only enabled when MutationEventsEnabled
is false.

Some tests are failing as specification updates and test changes are
still under review. This will be corrected in subsequent changes.

With all the settings in their current state (i.e., this is not
enabled) this tested as neutral on Speedometer.

Canonical link: https://commits.webkit.org/307064@main
